File: src/main/kotlin/reverseVowelsOfString.kt
//       date: 2022-11-04
//      title: Reverse Vowels of a String
//        url: https://leetcode.com/problems/reverse-vowels-of-a-string/
// difficulty: easy
//------------------------------------------------------------------------------

fun reverseVowels(s: String): String {
    val characters = s.toCharArray()

    var left = 0
    var right = characters.lastIndex

    while (left < right) {
        while (left < right && !isVowel(characters[left])) {
            left++
        }
        while (left < right && !isVowel(characters[right])) {
            right--
        }

        val tmp = characters[left]
        characters[left] = characters[right]
        characters[right] = tmp

        left++
        right--
    }

    return String(characters)
}

private fun isVowel(chr: Char): Boolean = chr in "AEIOUaeiou"

File: haskell/ReverseVowelsOfString.hs
--       date: 2022-11-04
--      title: Reverse Vowels of a String
--        url: https://leetcode.com/problems/reverse-vowels-of-a-string/
-- difficulty: easy
--------------------------------------------------------------------------------
module ReverseVowelsOfString
  ( reverseVowels
  , tests
  ) where

import           Test.HUnit

reverseVowels :: String -> String
reverseVowels str = replaceVowels vowels' str
  where
    vowels = filter isVowel str
    vowels' = reverse vowels

isVowel :: Char -> Bool
isVowel chr = chr `elem` "AEIOUaeiou"

replaceVowels :: String -> String -> String
replaceVowels [] ys = ys
replaceVowels _ [] = []
replaceVowels (x:xs) (y:ys) =
  if isVowel y
    then x : replaceVowels xs ys
    else y : replaceVowels (x : xs) ys

--------------------------------------------------------------------------------
-- Tests
tests :: Test
tests =
  TestList
    [ TestCase $ assertEqual "" "holle" $ reverseVowels "hello"
    , TestCase $ assertEqual "" "leotcede" $ reverseVowels "leetcode"
    , TestCase $
      assertEqual "" "zAUdrXMZFvBCYsUSTSonKtXiJdFoyzQM" $
      reverseVowels "zoidrXMZFvBCYsoSTSUnKtXUJdFAyzQM"
    , TestCase $
      assertEqual "" "HXxLFDKztPhYQhxHAxhKiTnWFdLJuamH" $
      reverseVowels "HXxLFDKztPhYQhxHaxhKuTnWFdLJiAmH"
    , TestCase $
      assertEqual "" "nBGWAfIkqNkvavDlrICzVUWqNjuSHKQz" $
      reverseVowels "nBGWufUkqNkvIvDlraCzVIWqNjASHKQz"
    ]
